Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -22,460 +22,653 @@
+				// convertIfcFaceList, convertIfcFace, convert3DPointsTo2D, triangulateFace, convertIfcCartesianPoint2DVector
+				//------------------------------------------------------------------------------------------------------------
+
+						/*! \brief  Converts \c IfcFace to a polygon and adds it to the carve PolyhedronData vector.
+						\param	faces \c IfcFace entity to be interpreted.
+						\param	pos
+						\param	itemData
+						\return polygon carve polygon
+						\note The \c IfcFaceList can be an open or closed shell.
+						*/
+
+				void convertIfcFaceList(const std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFace>>& faces,
+					const carve::math::Matrix& pos,
+					std::shared_ptr<ItemData> itemData) {
+
+					// Carve polygon of the converted face list
 					std::shared_ptr<carve::input::PolyhedronData> polygon(new carve::input::PolyhedronData());
-					// contains polygon indices of vertices (x,y,z converted to string)
+
+					// Contains polygon indices of vertices (x,y,z converted to string)
 					std::map<std::string, uint32_t> polygonIndices;
 
-					// loop through all faces
-					for(auto it = faces.cbegin(); it != faces.cend(); ++it) {
-						std::shared_ptr<typename IfcEntityTypesT::IfcFace> face = (*it);
+					// Loop through all faces
+					for (auto it = faces.cbegin(); it != faces.cend(); ++it) {
+						EXPRESSReference<typename IfcEntityTypesT::IfcFace> face = (*it);
 
-						if(!convertIfcFace(face, pos, polygon, polygonIndices)) {
+						if (!convertIfcFace(face, pos, polygon, polygonIndices)) {
 							std::stringstream text;
-							text << "IFC Face convertion failed with faces #" << faces.at(0)->getId() << "-" << faces.at(faces.size() - 1)->getId();
+							text << "IFC Face conversion failed with faces #" << faces.at(0)->getId() << "-" << faces.at(faces.size() - 1)->getId();
 
 							throw std::exception(text.str().c_str());
 						}
 					}
 
-					// IfcFaceList can be a closed or open shell,
-					// so let the calling function decide where to put it
-					item_data->open_or_closed_polyhedrons.push_back(polygon);
+					// IfcFaceList can be a closed or open shell, so let the calling function decide where to put it
+					itemData->open_or_closed_polyhedrons.push_back(polygon);
 				}
 
-				bool convertIfcFace(const std::shared_ptr<typename IfcEntityTypesT::IfcFace>& face,
+				/*! \brief  Converts \c IfcFace to a vector of face vertices.
+				\param	face \c IfcFace entity to be interpreted.
+				\param	pos
+				\param	polygon
+				\param	polygonIndices
+				\return conversionFailed true/false
+				\note	At the end, the calculated and merged face vertices are handed over to the \c triangulateFace function.
+				*/
+
+				bool convertIfcFace(const EXPRESSReference<typename IfcEntityTypesT::IfcFace>& face,
 					const carve::math::Matrix& pos,
 					std::shared_ptr<carve::input::PolyhedronData> polygon,
-					std::map<std::string, uint32_t>& polygonIndices) const throw(...)
-				{

```

Comment:
Please convert `pos` and `polygon` to return types and return them from the function.

Reply:
Does the same apply here @pjanck? Only return polygon?

Reply:
Yes.